Source/Core/Core/PowerPC/MMU.cpp

```diff
@@ -217,31 +250,71 @@ __forceinline void WriteToHardware(u32 em_address, const T data)
 			}
 			else
 			{
-				mmio_mapping->Write(em_address, data);
+				Memory::mmio_mapping->Write(em_address, data);
 				return;
 			}
 		}
-		else if (segment == 0x8 || segment == 0xC || segment == 0x0)
+		if ((segment == 0x8 || segment == 0xC) && (em_address & 0x0FFFFFFF) < Memory::REALRAM_SIZE)
 		{
-			*(T*)&m_pRAM[em_address & RAM_MASK] = bswap(data);
+			*(T*)&Memory::m_pRAM[em_address & 0x0FFFFFFF] = bswap(data);
 			return;
 		}
-		else if (m_pEXRAM && (segment == 0x9 || segment == 0xD || segment == 0x1))
+		if (Memory::m_pEXRAM && (segment == 0x9 || segment == 0xD) && (em_address & 0x0FFFFFFF) < Memory::EXRAM_SIZE)
 		{
-			*(T*)&m_pEXRAM[em_address & EXRAM_MASK] = bswap(data);
+			*(T*)&Memory::m_pEXRAM[em_address & 0x0FFFFFFF] = bswap(data);
 			return;
 		}
-		else if (segment == 0xE && (em_address < (0xE0000000 + L1_CACHE_SIZE)))
+		if (segment == 0xE && (em_address < (0xE0000000 + Memory::L1_CACHE_SIZE)))
 		{
-			*(T*)&m_pL1Cache[em_address & L1_CACHE_MASK] = bswap(data);
+			*(T*)&Memory::m_pL1Cache[em_address & 0x0FFFFFFF] = bswap(data);
 			return;
 		}
 	}
 
-	if (bFakeVMEM && (segment == 0x7 || segment == 0x4))
+	if (Memory::bFakeVMEM && performTranslation && (segment == 0x7 || segment == 0x4))
 	{
 		// fake VMEM
-		*(T*)&m_pFakeVMEM[em_address & FAKEVMEM_MASK] = bswap(data);
+		*(T*)&Memory::m_pFakeVMEM[em_address & Memory::FAKEVMEM_MASK] = bswap(data);
+		return;
+	}
+
+	if (!performTranslation)
```

Comment:
The PowerPC can't (and hence shouldn't attempt) write to MMIOs with data translation off. The reason being all data access in real mode is performed using WIMG access mode 0011 (write-back, caching enabled) which makes the writes go to the cache instead of directly to the hardware. At some point later the cache line will be evicted, sending 32 bytes across the relevant hardware address range... you can probably guess why this doesn't play nice with real hardware.
Unless there's a good reason to keep this block of code in I would suggest replacing it with an abort() equivalent.